File: src/main/java/com/now/nowbot/service/messageServiceImpl/GetIDService.java
package com.now.nowbot.service.messageServiceImpl;

import com.now.nowbot.config.Permission;
import com.now.nowbot.qq.event.MessageEvent;
import com.now.nowbot.service.MessageService;
import com.now.nowbot.service.osuApiService.OsuUserApiService;
import com.now.nowbot.throwable.GeneralTipsException;
import com.now.nowbot.util.DataUtil;
import com.now.nowbot.util.Instruction;
import jakarta.annotation.Resource;
import org.springframework.stereotype.Service;
import org.springframework.util.CollectionUtils;
import org.springframework.util.StringUtils;

import java.util.regex.Matcher;

@Service("GET_ID")
public class GetIDService implements MessageService<Matcher> {
    @Resource
    OsuUserApiService userApiService;

    @Override
    public boolean isHandle(MessageEvent event, String messageText, DataValue<Matcher> data) throws Throwable {
        var m = Instruction.GET_ID.matcher(messageText);
        if (m.find()) {
            data.setValue(m);
            return true;
        } else return false;
    }

    @Override
    public void HandleMessage(MessageEvent event, Matcher matcher) throws Throwable {
        var from = event.getSubject();

        if (Permission.isCommonUser(event)) {
            throw new GeneralTipsException(GeneralTipsException.Type.G_Permission_Group);
        }

        var names = DataUtil.splitString(matcher.group("data"));

        if (CollectionUtils.isEmpty(names)) throw new GeneralTipsException(GeneralTipsException.Type.G_Fetch_List);

        StringBuilder sb = new StringBuilder();

        for (var name : names) {
            if (! StringUtils.hasText(name)) {
                break;
            }

            long id;

            try {
                id = userApiService.getOsuId(name);
            } catch (Exception e) {
                sb.append("name=").append(name).append(" not found").append(',').append(' ');
                break;
            }

            sb.append(id).append(',').append(' ');
        }


        from.sendMessage(sb.substring(0, sb.length() - 2));
    }
}

File: src/main/java/com/now/nowbot/service/messageServiceImpl/GetIDService.kt
package com.now.nowbot.service.messageServiceImpl

import com.now.nowbot.config.Permission
import com.now.nowbot.dao.BindDao
import com.now.nowbot.qq.event.MessageEvent
import com.now.nowbot.service.MessageService
import com.now.nowbot.service.MessageService.DataValue
import com.now.nowbot.service.osuApiService.OsuUserApiService
import com.now.nowbot.throwable.GeneralTipsException
import com.now.nowbot.throwable.botRuntimeException.BindException
import com.now.nowbot.util.AsyncMethodExecutor
import com.now.nowbot.util.DataUtil.splitString
import com.now.nowbot.util.Instruction
import io.ktor.util.collections.*
import org.springframework.stereotype.Service
import kotlin.jvm.optionals.getOrNull

@Service("GET_ID")
class GetIDService(private val userApiService: OsuUserApiService, private val bindDao: BindDao) : MessageService<List<String>> {

    @Throws(Throwable::class) override fun isHandle(
        event: MessageEvent,
        messageText: String,
        data: DataValue<List<String>>
    ): Boolean {
        val m = Instruction.GET_ID.matcher(messageText)
        if (m.find()) {

            val str: String? = m.group("data")

            val names = if (event.isAt) {
                val b = bindDao.getQQLiteFromQQ(event.target).getOrNull() ?: throw BindException.NotBindException.UserNotBindException()

                event.reply(b.osuUser.osuID.toString())
                return false
            } else {
                splitString(str)
            }

            data.value = names ?: throw GeneralTipsException(GeneralTipsException.Type.G_Fetch_List)
            return true
        } else return false
    }

    @Throws(Throwable::class) override fun HandleMessage(event: MessageEvent, param: List<String>) {
        if (Permission.isCommonUser(event)) {
            throw GeneralTipsException(GeneralTipsException.Type.G_Permission_Group)
        }

        val sb = StringBuilder()

        val actions = param.map {
            return@map AsyncMethodExecutor.Supplier<Pair<String, Long>> {
                return@Supplier try {
                    it to userApiService.getOsuID(it)
                } catch (e: Exception) {
                    it to -1L
                }
            }
        }

        val ids = AsyncMethodExecutor.awaitSupplierExecute(actions)
            .filterNotNull()
            .filter { it.second > 0L }
            .toMap()

        param.forEach {
            val id = ids[it]

            sb.append(id).append(',')
        }

        /*

        for (name in names) {
            if (name.isBlank()) {
                continue
            }

            val id: Long

            try {
                id = userApiService.getOsuId(name)
            } catch (e: Exception) {
                sb.append("name=").append(name).append(" not found").append(',')
                continue
            }

            sb.append(id).append(',')
        }

         */

        event.reply(sb.toString().removeSuffix(","))
    }
}
